### ISSUE TITLE
Implicit arguments

### ISSUE THREAD

**Issue by robrix** [antitypical/Manifold#155]
This is the definition of `List.pure` added in #154:

``` swift
let pure = Declaration("List.pure",
    type: .Type => { A in A --> List.ref[A] },
    value: () => { A, a in cons[A, a, `nil`[A]] })
```

Syntax notwithstanding, this is much like in a “real” language. However, since there are no implicit arguments, we always have to apply the type parameters explicitly in the values of functions. This is (apparently) unnecessary! We should be able to reduce that down to:

``` swift
let pure = Declaration("List.pure",
    type: () => { A in A --> List.ref[A] },
    value: () => { a in cons[a, `nil`] })
```

Apparently, implicit arguments will allow this. I just have no idea how they work.

**Comment by robrix** [comment]
- The idris wiki has [a page on erasure annotations](https://github.com/idris-lang/Idris-dev/wiki/Egg-%232:-Erasure-annotations).
- Karl Crary, Stephanie Weirich, Greg Morrisett _[Intentional Polymorphism in Type-Erasure Semantics](http://www.cs.cornell.edu/talc/papers/typepass.pdf)_

**Comment by robrix** [comment]
Turns out I was conflating implicit arguments and type erasure. Filed #160 about erasure.

**Comment by robrix** [comment]
This was _not_ closed by #161.

**Comment by robrix** [comment]
Looks like it requires unification.
